pkg/apis/rollouts/v1alpha1/analysis_types.go

```diff
@@ -193,6 +195,40 @@ type JobMetric struct {
 	Spec     batchv1.JobSpec   `json:"spec" protobuf:"bytes,2,opt,name=spec"`
 }
 
+// CloudWatchMetric defines the cloudwatch query to perform canary analysis
+type CloudWatchMetric struct {
+	Interval          DurationString              `json:"interval,omitempty" protobuf:"bytes,1,opt,name=interval,casttype=DurationString"`
+	MetricDataQueries []CloudWatchMetricDataQuery `json:"metricDataQueries" protobuf:"bytes,2,opt,name=metricDataQueries"`
+}
+
+// CloudWatchMetricDataQuery defines the cloudwatch query
+type CloudWatchMetricDataQuery struct {
+	Id         *string               `json:"Id,omitempty" protobuf:"bytes,1,opt,name=Id"`
+	Expression *string               `json:"Expression,omitempty" protobuf:"bytes,2,opt,name=Expression"`
+	Label      *string               `json:"Label,omitempty" protobuf:"bytes,3,opt,name=Label"`
+	MetricStat *CloudWatchMetricStat `json:"MetricStat,omitempty" protobuf:"bytes,4,opt,name=MetricStat"`
+	Period     *int32                `json:"Period,omitempty" protobuf:"varint,5,opt,name=Period"`
+	ReturnData *bool                 `json:"ReturnData,omitempty" protobuf:"bytes,6,opt,name=ReturnData"`
+}
+
+type CloudWatchMetricStat struct {
+	Metric *CloudWatchMetricStatMetric `json:"Metric,omitempty" protobuf:"bytes,1,opt,name=Metric"`
+	Period *int32                      `json:"Period,omitempty" protobuf:"varint,2,opt,name=Period"`
+	Stat   *string                     `json:"Stat,omitempty" protobuf:"bytes,3,opt,name=Stat"`
+	Unit   string                      `json:"Unit,omitempty" protobuf:"bytes,4,opt,name=Unit"`
+}
+
+type CloudWatchMetricStatMetric struct {
+	Dimensions []CloudWatchMetricStatMetricDimension `json:"Dimensions,omitempty" protobuf:"bytes,1,opt,name=Dimensions"`
+	MetricName *string                               `json:"MetricName,omitempty" protobuf:"bytes,2,opt,name=MetricName"`
+	Namespace  *string                               `json:"Namespace,omitempty" protobuf:"bytes,3,opt,name=Namespace"`
+}
+
+type CloudWatchMetricStatMetricDimension struct {
+	Name  *string `json:"Name,omitempty" protobuf:"bytes,1,opt,name=Name"`
+	Value *string `json:"Value,omitempty" protobuf:"bytes,2,opt,name=Value"`
+}
+
```

Comment:
I like this new approach you use to use structured objects, instead of my suggestion of making `CloudWatchMetricDataQuery` being a string field. However, I think we'll need to modify this some more. I believe issues 1 and 2 may be because you copied exact datastructures from AWS go client and used as our own types:

1. There are now a lot of new "API rule violations" because of the casing of the json field name. Kubernetes APIs require names to be camel-cased always starting with lowercase letter, but here we use uppercase in the field names. e.g.:

```golang
Name  *string `json:"Name,omitempty" protobuf:"bytes,1,opt,name=Name"`
```

2. Some fields were made as pointers, when they should always have a value required. For example, I don't think these should be allowed to be pointers. Can we make everything non-pointers unless we need to explicitly distinguish between not supplied vs. empty?

```golang
type CloudWatchMetricStatMetricDimension struct {
	Name  *string `json:"Name,omitempty" protobuf:"bytes,1,opt,name=Name"`
	Value *string `json:"Value,omitempty" protobuf:"bytes,2,opt,name=Value"`
}
```

3. With this approach for using structured data (instead of a giant string field), we will not allow some things to be parameterized with Analysis arguments. For example, the `Period` field cannot be parameterized because it is an `int32`. 
```
	Period     *int32                `json:"Period,omitempty" protobuf:"varint,5,opt,name=Period"`
```

My question is: is `Period` something we should allow to be parameterized? If so, i think this should be made into an `intstrutil.IntOrString` instead of an `int32`

Reply:
1. I've fixed
2. I've fixed the pointer which isn't necessary I refer https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_MetricDataQuery.html
3. I've fixed for that I use `intstrutil.IntOrString` instead of `int32`